core/src/main/scala/org/apache/spark/sql/crossdata/catalog/CatalogChain.scala

```diff
@@ -44,20 +47,60 @@ object CatalogChain {
 private[crossdata] class CatalogChain private(val temporaryCatalogs: Seq[XDTemporaryCatalog],
                                               val persistentCatalogs: Seq[XDPersistentCatalog],
                                               val streamingCatalogs: Option[XDStreamingCatalog]
-                                               )(override val conf: CatalystConf) extends XDCatalog with SparkLoggerComponent {
+                                               )(implicit val xdContext: XDContext) extends XDCatalog with SparkLoggerComponent {
 
   import XDCatalogCommon._
 
+  override val conf: CatalystConf = xdContext.conf
+
   private val catalogs: Seq[XDCatalogCommon] = temporaryCatalogs ++: persistentCatalogs ++: streamingCatalogs.toSeq
 
   private implicit def crossdataTable2tableIdentifier(xdTable: CrossdataTable): TableIdentifier =
     TableIdentifier(xdTable.tableName, xdTable.dbName)
 
-  private def chainedLookup[R](lookup: XDCatalogCommon => Option[R]): Option[R] =
-    catalogs.view map lookup collectFirst {
-      case Some(res) => res
+  /**
+    * Apply the lookup function to each underlying catalog until a [[LogicalPlan]] is found. If the table is found in a
+    * temporary catalog, the relation is saved into the previous temporary catalogs.
+    */
+  private def chainedLookup(lookup: XDCatalogCommon => Option[LogicalPlan], tableIdentifier: TableIdentifier): Option[LogicalPlan] = {
+    val (relationOpt, previousCatalogs) = takeUntilRelationFound(lookup, temporaryCatalogs)
+
+    if (relationOpt.isDefined) {
+      previousCatalogs.foreach(_.saveTable(tableIdentifier, relationOpt.get)) //TODO tableIdentifier should be available
+      relationOpt
+    } else {
+      (persistentCatalogs ++: streamingCatalogs.toSeq).view map lookup collectFirst {
+        case Some(res) => res
+      }
     }
 
+  }
+
+
+  /**
+    * Apply the lookup function to each temporary catalog until a relation [[R]] is found. Returns the list of catalog,
+    * until a catalog satisfy the predicate 'lookup'.
+    *
+    * @param lookup lookup function
+    * @param tempCatalogs a seq of temporary catalogs
+    * @return a tuple (optionalRelation, firstNonMatchingLookupCatalogs)
+    */
+  private def takeUntilRelationFound[R](lookup: XDCatalogCommon => Option[R], tempCatalogs: Seq[XDTemporaryCatalog]): (Option[R], Seq[XDTemporaryCatalog]) = {
```

Comment:
I like monadic operations more, besides `:+` is O(n), thus the function O(n^2). In this case is not so important since the input length (n) will be not greater than 3 or 4. 

The following implementation is O(n) and uses monadic operations:

```scala
  private def takeUntilRelationFound2[R](lookup: XDCatalogCommon => Option[R], tempCatalogs: Seq[XDTemporaryCatalog]):
    (Option[R], Seq[XDTemporaryCatalog]) = {

    val (res: Option[R], idx: Int) = (tempCatalogs.view map(lookup) zipWithIndex) collectFirst {
      case e @ (res: Option[R @ unchecked], idx) if res.isDefined => e
    } getOrElse (None, 0)

    (res, tempCatalogs.take(idx+1))
  }
```